### PR TITLE
Add beta tester tool to force the coming soon landing pages to display on the front-end

### LINKED ISSUE #48655: Add beta tester tool to show coming soon landing page
## Issue Description

To make it easier to test the coming soon landing pages and flow, add a tool to WooCommerce Beta Tester to force the coming soon landing pages to show no matter what the site visibility settings are.

## Acceptance Criteria

* Tester can force the coming soon landing pages to display on the front-end, independently of the site visibility settings.
* Tester can select whether to show the "whole site" or "store only" template.

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/48655

Add a tool to WooCommerce Beta Tester to force the coming soon landing pages to show regardless of the site visibility settings.

Screenshot 2024-07-10 at 19 52 01

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch locally.
2. cd to `plugins/woocommerce-beta-tester` and run `pnpm run build:zip`
or using [woocommerce-beta-tester.zip](https://github.com/user-attachments/files/16188838/woocommerce-beta-tester.zip)
3. Create a new JN site with the latest WooCommerce 
4. Upload and activate `woocommerce-beta-tester.zip`
5. Go to `/wp-admin/admin.php?page=wc-settings&tab=site-visibility`
6. Update the site visibility to `Live`
7. Go to `Tools -> WCA Test Helper -> Tools`
8. Select `Whole Site` or `Store Only` and click on `Run
9. Confirm it ran successfully
10. Open an incognito window and visit the site front end
11. Confirm the coming soon landing page is shown based on the mode selected in step 8
12. Go to Site Visibility Settings and confirm the site visibility is still set to `Live`
13. Go to task list and confirm the LYS task is marked as complete
14. Play around with the site and confirm everything works as expected

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by moon0326 — APPROVED** [review]
Nice addition! LGTM and tested as expected 👍 🚀

File: plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-override-coming-soon-options.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+			if ('site' === $mode || 'store' === $mode) {
+				return 'yes';
+			}
+			return $value;
+		});
+	}
+
+	private function override_woocommerce_store_pages_only_option($mode) {
+		add_filter('option_woocommerce_store_pages_only', function ($value) use ($mode) {
+			if (is_admin() || wp_doing_ajax()) {
```
Comment (ilyasfoo) [inline plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-override-coming-soon-options.php]: Should we also add `defined('DOING_CRON') && DOING_CRON`?

File: plugins/woocommerce-beta-tester/src/tools/data/resolvers.js [hunk 2]
```diff
@@ -85,3 +86,17 @@ export function* getBlockTemplateLoggingThreshold() {
...
+
+export function* getComingSoonMode() {
+	const path = `${ API_NAMESPACE }/tools/get-force-coming-soon-mode/v1`;
+
+	try {
+		const response = yield apiFetch( {
+			path,
+			method: 'GET',
+		} );
+		yield setComingSoonMode( response );
```
Comment (ilyasfoo) [inline plugins/woocommerce-beta-tester/src/tools/data/resolvers.js]: I suspect this might not set the correct value? Do we need to parse the response?

**Review by ilyasfoo — CHANGES_REQUESTED** [review]
Couple of issues I found:

1. Post-loading, `comingSoonMode` doesn't seem to hold the correct value, I suspect the resolver might not set the correct value:

https://github.com/user-attachments/assets/f8c7d9a6-1586-4ca0-a0e0-6c780b1063b0

1. The store-only mode did not work, it shows the entire site page instead:



https://github.com/user-attachments/assets/35b1e8ae-ecb3-44d4-acc9-968a88d33969

File: plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-override-coming-soon-options.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-override-coming-soon-options.php]: Nope, actually, we don't need this check. It's not necessary as we already check if request is from frontend before calling the function.

File: plugins/woocommerce-beta-tester/src/tools/data/resolvers.js (hunk 2 above)
Comment (chihsuan) [reply plugins/woocommerce-beta-tester/src/tools/data/resolvers.js]: We don't need to parse the response. The response should be a string. We can directly use it.

 I think I might build a wrong zip file. 😓  It's working fine locally. I'll test on JN and let you know.

**Comment by chihsuan** [comment]
@ilyasfoo Thanks for caching the issues. 🙏  I tested it on JN and found that one of the issues is that the cache is not getting invalidated when the setting is changed. I've fixed that by calling coming soon's invalidate_caches() method in 
5670498df3221b83c9d040251db1514b8f28d488. Everything should be working fine now. Please have a look at it again. Thanks!

File: plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-override-coming-soon-options.php (hunk 1 above)
Comment (chihsuan) [reply plugins/woocommerce-beta-tester/includes/class-wc-beta-tester-override-coming-soon-options.php]: Removed it and fixed lint issues in 130739441f9d61a84029a5095c37b4ba072767d1

**Review by ilyasfoo — COMMENTED** [review]
@chihsuan I'm still having trouble with both issues (JSON error at tool & entire site coming soon) 😅 

I've built from my local with the latest changes from PR, as I inspect the code in my JN:

image


https://github.com/user-attachments/assets/68540bfc-7cbe-4a73-a66b-e902fd38b3ba

I can pass you the JN instance via DM, could you take a look? 🙏

**Comment by chihsuan** [comment]
Thanks for providing the JN instance, @ilyasfoo. I was able to reproduce the issue and found that `Run version update callbacks` and `Update block template logging threshold` tools were also affected by the same problem.  The root cause is that we didn't set the default values for the states. I've fixed them all together in f204ee5f490bc8b60a7f250f8de59362ea0c5194

Please have another look at the code and let me know if it works for you now.  🙏 

Before:


https://github.com/user-attachments/assets/b19c9e08-ae8f-420f-af1a-0b3f42c5e89c


After:

https://github.com/user-attachments/assets/cfa3fe2c-6b13-467d-b0eb-50a446ffb016

File: plugins/woocommerce-beta-tester/api/tools/trigger-update-callbacks.php [hunk 3]
```diff
@@ -48,7 +48,7 @@ function trigger_selected_update_callbacks( $request ) {
 	$update_callbacks = $db_updates[ $version ];
 
 	foreach ( $update_callbacks as $update_callback ) {
-		call_user_func( $update_callback );
+		\WC_Install::run_update_callback( $update_callback );
```
Comment (chihsuan) [inline plugins/woocommerce-beta-tester/api/tools/trigger-update-callbacks.php:51]: `call_user_func( $update_callback );` caused a fatal error when the callback was not loaded. `	\WC_Install::run_update_callback` ensures the callback is loaded before calling it.

https://github.com/woocommerce/woocommerce/blob/4d68cd486e07b3dabffe307157ddfe3737eb39c5/plugins/woocommerce/includes/class-wc-install.php#L392-L400

**Review by ilyasfoo — COMMENTED** [review]
Thanks, @chihsuan! The fix for the fatal tests well for me, but there's still the issue with store-only mode, at least in my JN 😅 . I think the `init` action does not run when the [line here](https://github.com/woocommerce/woocommerce/blob/b095c64b580952508dc28208f89756d088577871/plugins/woocommerce/patterns/coming-soon.php#L12) is executed.

Here's how I reproduced it:

1. Create a fresh JN with this branch and install `woocommerce-beta-tester` built from this branch
2. Skip OBW
3. Set store to live
4. Set the `Force coming soon page to show` mode to `Store Only`
5. Browse cart page with incognito and observe it does not display entire site coming soon template

**Comment by chihsuan** [comment]
Thanks @ilyasfoo, I was able to reproduce the issue using Live branch and the steps you provided. I was using the latest release WooCommerce plugin, so I didn't see the issue. It's indeed related to the `init` action. Fixed it in b6604574276d373d95f33cfd2746ab9f00cb7430.

**Review by ilyasfoo — APPROVED** [review]
Awesome, I confirm it's now fixed! Thanks @chihsuan  🚢

**Pull request merged**